wire-ios-sync-engine/Source/Repositories/ConnectionsRepositiory.swift

```diff
@@ -0,0 +1,146 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+import WireAPI
+import WireDataModel
+
+/// Facilitate access to connections related domain objects.
+///
+/// A repository provides an abstraction for the access and storage
+/// of domain models, concealing how and where the models are stored
+/// as well as the possible source(s) of the models.
+protocol ConnectionsRepositoryProtocol {
+
+    /// Pull self team metadata frmo the server and store locally.
+
+    func pullConnections() async throws
+}
+
+final class ConnectionsRepository: ConnectionsRepositoryProtocol {
+
+    private let connectionsAPI: any ConnectionsAPI
+    private let context: NSManagedObjectContext
+
+    init(
+        connectionsAPI: any ConnectionsAPI,
+        context: NSManagedObjectContext
+    ) {
+        self.connectionsAPI = connectionsAPI
+        self.context = context
+    }
+
+    /// Retrieve from backend and store connections location
+
+    public func pullConnections() async throws {
+        let connectionsPager = try await connectionsAPI.getConnections()
+
+        for try await connections in connectionsPager {
+            await withThrowingTaskGroup(of: Void.self) { taskGroup in
+                for connection in connections {
+                    taskGroup.addTask {
+                        try await self.storeConnection(connection)
+                    }
+                }
+            }
+        }
+    }
+
+    /// Save connection and related objects to local storage.
+    /// - Parameter connectionPayload: connection object from WireAPI
+
+    private func storeConnection(_ connectionPayload: Connection) async throws {
+        try await context.perform { [self] in
+
+            let connection = try storedConnection(from: connectionPayload)
+
+            let conversation = try storedConversation(from: connectionPayload, with: connection)
+
+            connection.to.oneOnOneConversation = conversation
+
+            try context.save()
+        }
+    }
+
+    /// Create or update conversation related to the connection's sender
+    /// - Parameters:
+    ///   - connection: connection payload from WireAPI
+    ///   - storedConnection: ZMConnection object stored locally
+    /// - Returns: conversation object stored locally
+
+    private func storedConversation(from connection: Connection, with storedConnection: ZMConnection) throws -> ZMConversation {
+
+        guard let conversationID = connection.conversationId ?? connection.qualifiedConversationId?.uuid else {
+            throw ConnectionsRepositoryError.missingConversationId
+        }
+
+        let conversation = ZMConversation.fetchOrCreate(
+            with: conversationID,
+            domain: connection.qualifiedConversationId?.domain,
```

Comment:
It will search the local domain. Other platforms now always use domains, unfortunately for us we still allow optional domains because of some external requirements :(